fixes/data_submission_recids.py

```diff
@@ -0,0 +1,134 @@
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+from datetime import datetime
+import logging
+
+from flask import current_app
+from flask.cli import with_appcontext
+from invenio_db import db
+from sqlalchemy import func
+
+from hepdata.cli import fix
+from hepdata.ext.elasticsearch.api import index_record_ids, push_data_keywords
+from hepdata.modules.records.utils.common import get_record_by_id
+from hepdata.modules.records.utils.doi_minter import generate_dois_for_submission
+from hepdata.modules.records.utils.submission import finalise_datasubmission
+from hepdata.modules.submission.models import HEPSubmission, DataSubmission
+
+logging.basicConfig()
+log = logging.getLogger('data_submission_recids')
+log.setLevel('INFO')
+
+
+@fix.command()
+@with_appcontext
+def fix_data_submission_recids():
+    """Update associated_recids for data submissions with version > 1"""
+    submissions_to_update = HEPSubmission.query \
+        .filter(HEPSubmission.version > 1) \
+        .filter(HEPSubmission.overall_status == 'finished') \
+        .order_by(HEPSubmission.publication_recid, HEPSubmission.version).all()
+
+    log.info(f"Updating data submissions for {len(submissions_to_update)} publications with version > 1")
+
+    for i, hep_submission in enumerate(submissions_to_update):
+        # We have ordered the results so if this is the last entry in
+        # submissions_to_update, or the next entry has a different
+        # publication_recid, then this must be the latest version
+        is_latest = (i + 1 == len(submissions_to_update)) or \
+            submissions_to_update[i+1].publication_recid != hep_submission.publication_recid
+        _create_new_data_records(hep_submission, is_latest)
+
+    log.info("Checking for any more duplicate associated rec_ids...")
+    duplicates = DataSubmission.query.with_entities(
+        DataSubmission.associated_recid,
+        func.count(DataSubmission.associated_recid)) \
+        .filter(DataSubmission.associated_recid.isnot(None)) \
+        .group_by(DataSubmission.associated_recid) \
+        .having(func.count(DataSubmission.associated_recid) > 1).all()
+
+    if duplicates:
+        log.warning("There are still some records with duplicate associated_recids:")
+        for d in duplicates:
+            log.warning(f"    associated_recid: {d[0]}, count {d[1]}")
+    else:
+        log.info('No duplicates found 👍')
+
+
+def _create_new_data_records(hep_submission, is_latest=True):
+    log.info("Checking records for publication rec_id "
+             f"{hep_submission.publication_recid}, "
+             f"version {hep_submission.version} "
+             f"(inspire id {hep_submission.inspire_id})")
+
+    data_submissions = DataSubmission.query.filter_by(
+        publication_recid=hep_submission.publication_recid,
+        version=hep_submission.version
+    )
```

Comment:
Can you please add `.order_by(DataSubmission.id.asc())` so that the new `associated_recid` values will be assigned in the same order as the data tables in a record?

Reply:
Should we also do the same in [`do_finalise`](https://github.com/HEPData/hepdata/blob/ae539212fa7f8f9a29803c9de85e3ff2def67707/hepdata/modules/records/utils/submission.py#L686)?

Reply:
Yes, please.  I had thought the `associated_recid` values were already assigned in the same order as data tables for a given record version, but checking my local database it seems to be the case for some records but not others, so better to force the same ordering.